rmw_fastrtps_cpp/src/listener_thread.cpp

```diff
@@ -0,0 +1,169 @@
+  if (RMW_RET_OK != rmw_trigger_guard_condition(common_context->listener_thread_gc)) {
+    goto fail;
+  }
+  try {
+    common_context->listener_thread.join();
+  } catch (...) {
+    goto fail;
+  }
+  if (RMW_RET_OK != rmw_destroy_guard_condition(common_context->listener_thread_gc)) {
+    goto fail;
+  }
+  return RMW_RET_OK;
+fail:
+  return RMW_RET_ERROR;
+}
+
+static
+void
+terminate(const char * error_message)
+{
+  RCUTILS_LOG_ERROR_NAMED(log_tag, "%s, terminating ...", error_message);
+  std::terminate();
+}
+
+void
+node_listener(rmw_context_t * context)
+{
+  assert(nullptr != context);
+  assert(nullptr != context->impl);
+  assert(nullptr != context->impl->common);
+  auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+  while (common_context->thread_is_running.load()) {
+    assert(nullptr != common_context->sub);
+    assert(nullptr != common_context->sub->data);
+    void * subscriptions_buffer[] = {common_context->sub->data};
+    void * guard_conditions_buffer[] = {common_context->listener_thread_gc->data};
+    rmw_subscriptions_t subscriptions;
+    rmw_guard_conditions_t guard_conditions;
+    subscriptions.subscriber_count = 1;
+    subscriptions.subscribers = subscriptions_buffer;
+    guard_conditions.guard_condition_count = 1;
+    guard_conditions.guard_conditions = guard_conditions_buffer;
+    // number of conditions of a subscription is 2
+    rmw_wait_set_t * wait_set = rmw_create_wait_set(context, 2);
+    if (nullptr == wait_set) {
+      terminate("failed to create wait set");
+    }
+    if (RMW_RET_OK != rmw_wait(
+        &subscriptions,
+        &guard_conditions,
+        nullptr,
+        nullptr,
+        nullptr,
+        wait_set,
+        nullptr))
+    {
+      terminate("rmw_wait failed");
+    }
+    if (subscriptions_buffer[0]) {
+      rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+      bool taken;
+      if (RMW_RET_OK != rmw_take(
+          common_context->sub,
+          static_cast<void *>(&msg),
+          &taken,
+          nullptr))
+      {
+        terminate("rmw_take failed");
+      }
+      if (taken) {
+        // TODO(ivanpauno): Should the program be terminated if taken is false?
+        if (std::strncmp(
+            reinterpret_cast<char *>(common_context->gid.data),
+            reinterpret_cast<char *>(&msg.gid.data),
+            RMW_GID_STORAGE_SIZE) == 0)
+        {
+          // ignore local messages
+          continue;
+        }
+        common_context->graph_cache.update_participant_entities(msg);
+        if (rcutils_logging_logger_is_enabled_for("rmw_dds_common",
+          RCUTILS_LOG_SEVERITY_DEBUG))
+        {
+          std::ostringstream ss;
+          ss << common_context->graph_cache;
+          RCUTILS_LOG_DEBUG_NAMED("rmw_fastrtps_cpp", "%s", ss.str().c_str());
+        }
+      }
+    }
+    if (RMW_RET_OK != rmw_destroy_wait_set(wait_set)) {
+      terminate("failed to destroy wait set");
+    }
+  }
+}
```

Comment:
I mean, the other error conditions in this function use terminate, so maybe doing something different is out of scope?

Reply:
There's not much to do when this thread fails, though it's true that the "correct thing" to do is shutting down the context corresponding to this listener thread, and not terminating everything.
But currently, shutting down a context is not trivial from `rmw` (adding a interrupt guard condition to the context would be a good solution).

Another solution is to just log an error instead of terminating.
If this thread stops working, the graph is not going to be correctly updated, but other things should still work correctly.

Reply:
It's fine to `std::terminate` for now, let's just log enough for the user to know why his/her process aborted. 

Ideally, yes, I think it'd be great for this thread to just shutdown the context it belongs to, but as @wjwwood says it's out of scope for this PR.